flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/source/split/BigQuerySourceSplitAssigner.java

```diff
@@ -0,0 +1,247 @@
+                                .setMaxStreamCount(readOptions.getMaxStreamCount());
+
+                // request the session
+                ReadSession session = client.createReadSession(builder.build());
+                LOG.info(
+                        "BigQuery Storage Read session, name: {},"
+                                + " estimated row count {}, estimated scanned bytes {},"
+                                + " streams count {}, expired time {} (seconds after epoch).",
+                        session.getName(),
+                        session.getEstimatedRowCount(),
+                        session.getEstimatedTotalBytesScanned(),
+                        session.getStreamsCount(),
+                        session.getExpireTime().getSeconds());
+                // get all the stream names added to the initialized state
+                remainingTableStreams.addAll(
+                        session.getStreamsList().stream()
+                                .map(stream -> stream.getName())
+                                .collect(Collectors.toList()));
+                initialized = true;
+            } catch (IOException ex) {
+                throw new RuntimeException(
+                        "Problems creating the BigQuery Storage Read session.", ex);
+            }
+        }
+    }
+
+    public void addSplitsBack(List<BigQuerySourceSplit> splits) {
+        for (BigQuerySourceSplit split : splits) {
+            remainingSourceSplits.add((BigQuerySourceSplit) split);
+            // we should remove the add-backed splits from the assigned list,
+            // because they are failed
+            assignedSourceSplits.remove(split.splitId());
+        }
+    }
+
+    public BigQuerySourceEnumState snapshotState(long checkpointId) {
+        return new BigQuerySourceEnumState(
+                Lists.newArrayList(remainingTableStreams),
+                alreadyProcessedTableStreams,
+                Lists.newArrayList(remainingSourceSplits),
+                assignedSourceSplits,
+                initialized);
+    }
+
+    public void close() {
+        // so far not much to be done here
+        LOG.info("BigQuery source split assigner is closed.");
+    }
+
+    public Optional<BigQuerySourceSplit> getNext() {
+        if (!remainingSourceSplits.isEmpty()) {
+            // return remaining splits firstly
+            BigQuerySourceSplit split = remainingSourceSplits.poll();
+            assignedSourceSplits.put(split.splitId(), split);
+            return Optional.of(split);
+        } else {
+            // it's turn for next collection
+            String nextStream = remainingTableStreams.poll();
+            if (nextStream != null) {
+                BigQuerySourceSplit split = new BigQuerySourceSplit(nextStream);
+                remainingSourceSplits.add(split);
+                alreadyProcessedTableStreams.add(nextStream);
+                return getNext();
```

Comment:
What's the expected depth of recursive calls?

Reply:
It shouldn't be more than 1.